src/instructlab/process/process.py

```diff
@@ -0,0 +1,178 @@
+# Standard
+from datetime import datetime
+import json
+import logging
+import os
+import subprocess
+import sys
+import time
+import uuid
+
+# Third Party
+import psutil
+
+# First Party
+from instructlab.configuration import DEFAULTS
+from instructlab.defaults import ILAB_PROCESS_MODES, ILAB_PROCESS_TYPES
+
+logger = logging.getLogger(__name__)
+
+
+def load_registry():
+    process_registry = {}
+    """Load the process registry from a file, if it exists."""
+    # we do not want a persistent registry in memory. This causes issues when in scenarios where you switch registry files (ex, in a unit test, or with multuple users)
+    # but the registry with incorrect processes still exists in memory.
+    if os.path.exists(DEFAULTS.REGISTRY_FILE):
+        with open(DEFAULTS.REGISTRY_FILE, "r") as f:
+            data = json.load(f)
+            for key, value in data.items():
+                process_registry[key] = value
+    else:
+        logger.info("No existing process registry found. Starting fresh.")
+    return process_registry
+
+
+def save_registry(process_registry):
+    """Save the current process registry to a file."""
+    with open(DEFAULTS.REGISTRY_FILE, "w") as f:
+        json.dump(dict(process_registry), f)
+
+
+class Tee:
+    def __init__(self, log_file):
+        """
+        Initialize a Tee object.
+
+        Args:
+            log_file (str): Path to the log file where the output should be written.
+        """
+        self.log_file = log_file
+        self.terminal = sys.stdout
+        self.log = log_file  # Line-buffered
+
+    def write(self, message):
+        """
+        Write the message to both the terminal and the log file.
+
+        Args:
+            message (str): The message to write.
+        """
+        self.terminal.write(message)
+        self.log.write(message)
+
+    def flush(self):
+        """
+        Ensure all data is written to the terminal and the log file.
+        """
+        self.terminal.flush()
+        self.log.flush()
+
+    def close(self):
+        """
+        Close the log file.
+        """
+        if self.log:
+            self.log.close()
+
+
+def add_process(
```

Comment:
This is a huge function, I think this should be its own class too, or broken down into smaller bits.

```
class ProcessManager:
    def __init__(self, process_registry, log_dir):
        self.process_registry = process_registry
        self.log_dir = log_dir

    def add_process(self, process_mode, process_type, target, extra_imports, **kwargs):
        # ... (implementation details)

    def _prepare_command(self, target, extra_imports, kwargs):
        # ... (command setup logic)

    def _setup_logging(self, process_type):
        # ... (logging setup logic)

    def _launch_process(self, cmd, log_file):
        # ... (process launching logic)

    def _capture_child_pids(self, parent_pid):
        # ... (child PID capturing logic)

    def _update_registry(self, pid, process_type, start_time, log_file, children_pids):
        # ... (registry update logic)
```